tap_nice_incontact/streams.py

```diff
@@ -604,6 +609,200 @@ def get_records(self,
+        :param stream_metadata: A dictionnary containing stream metadata
+        :param config: A dictionary containing tap config data
+        :param transformer: A singer Transformer object
+        :return: State data in the form of a dictionary
+        """
+        
+
+        start_date = singer.get_bookmark(state,
+                                        self.tap_stream_id,
+                                        self.replication_key,
+                                        config['start_date'])
+        bookmark_datetime = singer.utils.strptime_to_utc(start_date)
+        max_datetime = bookmark_datetime        
+        
+        with metrics.record_counter(self.tap_stream_id) as counter:
+            for record in self.get_records(config, bookmark_datetime):
+                # Convert data keys for CSV outputs to the correct field format
+                record = convert_data_keys(record)
+
+                if self.convert_data_types:
+                    record = convert_data_types(record, stream_schema)
+
+                transformed_record = transformer.transform(record, stream_schema, stream_metadata)
+                # pylint: disable=line-too-long
+                record_datetime = singer.utils.strptime_to_utc(transformed_record[self.replication_key])
+                if record_datetime >= bookmark_datetime:
+                    singer.write_record(self.tap_stream_id, transformed_record)
+                    counter.increment()
+                    max_datetime = max(record_datetime, bookmark_datetime)
+
+            bookmark_date = singer.utils.strftime(max_datetime)
+
+        state = singer.write_bookmark(state,
+                                    self.tap_stream_id,
+                                    self.replication_key,
+                                    bookmark_date)
+        singer.write_state(state)
+        return state
+
+    def get_records(self,
+                    config: dict = None,
+                    bookmark_datetime: datetime = None,
+                    is_parent: bool = False) -> Iterator:
+        if config.get('periods'):
+            period = config.get('periods', {}).get(self.tap_stream_id)
+        else:
+            period = self.default_period
+
+        for start, end in self.generate_date_range(bookmark_datetime, period=period):
+            start_date = singer.utils.strptime_to_utc(start)
+            end_date = singer.utils.strptime_to_utc(end)
+
+            try:
+                results = self.fetch_data_export(start_date, end_date, config)
+            except (NiceInContact403Exception, NiceInContact5xxException, ValueError) as e:
+                # We get a 403 exception if we are rate-limited by the data
+                # extraction API. Exclude these errors and wait until the
+                # next tap run to continue from the start datetime.
+                #
+                # Otherwise break if we receive a server error or invalid job status.
+                break
```

Comment:
I'm consistently getting retry attempts up to the limit since the default start time is 1s. Should we change the backoff args?

Reply:
I also think this is a pretty optimistic assumption that future runs will pick up the entire slack of the previous runs. This might produce some?

Reply:
I'm not sure whats a good way of handling this issue. We could wait like a minute or even longer for each period so that we don't request immediately after a success..?

Reply:
I've increased `MAX_RETRY` to 8 attempts, which should give us ~4min backoff in the last retry. I used this new backoff and it appears that the tap succeeds typically on the 6th try.

Generally running this tap multiple times in a row, it seems like it needs 2-5min in between job creation in order to successfully queue.

Reply:
With the updated retry attempts, I am now able to run it for multiple weeks worth of data (12/5-12/25) without getting stopped. Tap is still going -- so we should be able to successfully use this retry count for Ancestry in production

Reply:
The run completed successfully from 12/5 to today (1/9 midnight) so we should be good using this retry count 👍

Potentially this is something we can consider adding to the config as well, if we feel like we might need to tune this for different customers